qiskit/assembler/assemble_circuits.py

```diff
@@ -105,18 +112,76 @@ def _assemble_circuit(circuit):
             del instruction._condition
 
         instructions.append(instruction)
-    return QasmQobjExperiment(instructions=instructions, header=header,
-                              config=config)
+    return (QasmQobjExperiment(instructions=instructions, header=header, config=config),
+            pulse_library)
+
+
+def _assemble_pulse_gates(circuit, run_config):
+    """Assemble and return the circuit calibrations and associated pulse library, if there are any.
+    The calibrations themselves may reference the pulse library which is returned as a dict.
+
+    Args:
+        circuit (QuantumCircuit): circuit which may have pulse calibrations
+        run_config (RunConfig): configuration of the runtime environment
+
+    Returns:
+        tuple(QasmExperimentCalibrations, dict(str, array)): The calibrations and pulse library.
+    """
+    if not circuit.calibrations:
+        return None, None
+    if not hasattr(run_config, 'parametric_pulses'):
+        run_config.parametric_pulses = []
+    calibrations = []
+    pulse_library = {}
+    for gate, cals in circuit.calibrations.items():
+        for (qubits, params), schedule in cals.items():
+            qobj_instructions, _ = assemble_schedule(
+                schedule,
+                converters.InstructionToQobjConverter(PulseQobjInstruction),
+                run_config,
+                pulse_library)
+            calibrations.append(
+                GateCalibration(str(gate), list(qubits), list(params), qobj_instructions))
+    return QasmExperimentCalibrations(gates=calibrations), pulse_library
+
+
+def _extract_common_calibrations(experiments):
+    """Given a list of ``QasmQobjExperiment``s, each of which may have calibrations in their
+    ``config``, collect common calibrations into a global ``QasmExperimentCalibrations``
+    and delete them from their local experiments.
+    """
+    if not (experiments and all(hasattr(exp.config, 'calibrations') for exp in experiments)):
+        return experiments, None
+
+    common_calibrations = []
+    from collections import defaultdict
+    gate_hash_map = defaultdict(list)
+    for exp_idx, exp in enumerate(experiments):
+        for gate_idx, gate_cal in enumerate(exp.config.calibrations.gates):
+            gate_hash_map[str(gate_cal.to_dict())].append((exp_idx, gate_idx, gate_cal))
```

Comment:
Would it make more sense to do this mapping at the level of pulse schedules where the notion of equality and a hash already exist?